### ISSUE TITLE
Validation of Enum Field with Decimal Fails

### ISSUE THREAD

**Issue by jroy-sophos** [pydantic/pydantic#9559]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Similar to [this issue](https://github.com/pydantic/pydantic/issues/9375), we are seeing problems when validating Enum fields from the "Decimal" type. 

This previously worked in 2.4.2, but is no longer working with 2.7.2.

### Example Code

```Python
from enum import Enum
from decimal import Decimal

from pydantic import BaseModel

class EnumClass(Enum):
    enum_value = 1

class ModelClass(BaseModel):
    enum_field: EnumClass

validated_class = ModelClass.model_validate({"enum_field": Decimal(1)})
print(validated_class.enum_field)

# 2.4.2 - Prints "EnumClass.enum_value"
# 2.7.2 - Returns Validation Error: 
    # pydantic_core._pydantic_core.ValidationError: 1 validation error for ModelClass
    # enum_field
    # Input should be 1 [type=enum, input_value=Decimal('1'), input_type=Decimal]
    #     For further information visit https://errors.pydantic.dev/2.7/v/enum
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
        pydantic-core version: 2.10.1
          pydantic-core build: profile=release pgo=true
                 install path: {removed}/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
                     platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.8.0

             pydantic version: 2.7.2
        pydantic-core version: 2.18.3
          pydantic-core build: profile=release pgo=true
                 install path: {removed}/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
                     platform: Linux-5.15.0-107-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.8.0
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@jroy-sophos,

Thanks for reporting this. Looks like probably an issue related to our `enum` validation changes in 2.7.

I'll work on a fix for this in 2.8!

**Comment by sydney-runkle** [comment]
Might even get a fix out in a patch release - will keep you updated. Likely an issue in `pydantic-core`.

**Comment by mikeleppane** [comment]
Hey,

The Decimal type lacks support here:
https://github.com/pydantic/pydantic-core/blob/34d789fc510810eee507347d5e0897d8fb9045bb/src/validators/enum_.rs#L166

I might make a PR out of this during the weekend.

**Comment by sydney-runkle** [comment]
@mikeleppane,

Great, that would be awesome!

**Comment by davidhewitt** [comment]
I think it's not just related to `Decimal`, but any type which can compare equal with the enum members. For example, here's a custom class which also validated successfully on 2.4.2 but fails on 2.7.2:

```
from enum import Enum

from pydantic import BaseModel

class EnumClass(Enum):
    enum_value = 1

class ModelClass(BaseModel):
    enum_field: EnumClass

class IntWrappable:
    def __init__(self, value: int):
        self.value = value
    
    def __eq__(self, value: object) -> bool:
        return self.value == value

print(EnumClass(IntWrappable(1)))

ModelClass(enum_field=IntWrappable(1))
```

**Comment by jroy-sophos** [comment]
Hello,

Thank you very much for looking into this! I greatly appreciate the effort that has been made around making a general solution for this and other similar problems, which does seem for the best. 

Just for my understanding, how is this progressing? Is this a planned feature in an upcoming release? 

Thank you!

**Comment by sydney-runkle** [comment]
Hey @jroy-sophos,

Thanks for your inquiry! I believe we'll be working on some enum validator fixes in v2.10 in September. I'll keep you updated on this issue!
